core/src/main/java/io/atomix/core/election/impl/DefaultLeaderElectorService.java

```diff
@@ -187,6 +187,27 @@ public boolean promote(String topic, byte[] id) {
     }
   }
 
+  @Override
+  public boolean demote(String topic, byte[] id) {
+    try {
+      Leadership<byte[]> oldLeadership = leadership(topic);
+      if (oldLeadership == null ||
+              oldLeadership.candidates().stream().noneMatch(candidate -> Arrays.equals(candidate, id)) ||
+              Arrays.equals(oldLeadership.leader().id(), id)) {
+        return false;
+      }
+      elections.computeIfPresent(topic, (k, v) -> v.demote(id));
+      Leadership<byte[]> newLeadership = leadership(topic);
```

Comment:
Is it possible to figure out whether the operation is successful by looking at the `computeIfPresent` return value? Something like:
```
ElectorState newState = elections.computeIfPresent(topic, (k, v) -> v.demote(id));
if (newState != null && !Objects.equal(oldLeadership, newState.leader())) {
  // notify...
}
```

Reply:
I can but I don't have directly the leadership object I want to compare and I still need other code (what `leadership` function does under the hood). I believe it is fine - The service (IIRC) is executed by a single thread we should not worry about race